bin/src/hickory-dns.rs

```diff
@@ -229,10 +239,22 @@ async fn async_run(args: Cli) -> Result<(), String> {
 
     #[cfg(feature = "prometheus-metrics")]
     if !args.disable_prometheus && !config.disable_prometheus() {
-        let socket = args
+        let mut socket = args
             .prometheus_listen_addr
             .unwrap_or(config.prometheus_listen_addr());
 
+        // bind to select a port, then drop the port so it can be used
+        // needed for dynamic port selection (port 0) as it is currently not possible to
+        // obtain the selected port from PrometheusBuilder after it binds the SocketAddr
+        // during PrometheusBuilder::install() or PrometheusBuilder::build()
+        if socket.port() == 0 {
+            let listener = std::net::TcpListener::bind(socket)
```

Comment:
Is there an issue filed on the Prometheus repo to allow for either the port to be discoverable or for the TcpListener to be passed in?

Reply:
that seems pretty heavy, but ok.

Reply:
I filed #2947 for this follow-up work.